Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/TapToPlace/TapToPlace.cs

```diff
@@ -0,0 +1,467 @@
+            var transform = SolverHandler.TransformTarget;
+
+            Vector3 origin = transform.position;
+            Vector3 endpoint = transform.position + transform.forward;
+            currentRay.UpdateRayStep(ref origin, ref endpoint);
+
+            // Check if the current ray hits a magnetic surface
+            DidHit = MixedRealityRaycaster.RaycastSimplePhysicsStep(currentRay, MaxRaycastDistance, MagneticSurfaces, false, out currentHit);  
+        }
+
+        protected virtual void SetPosition()
+        {
+            // Change the position of the GameObjectToPlace if there was a hit, if not then place the object at the default distance
+            // relative to the TrackedTargetType position
+            
+            if (DidHit)
+            {
+                // take the current hit point and add an offset relative to the surface to avoid half of the object in the surface
+                GoalPosition = currentHit.point;  
+                AddOffset(currentHit.normal * SurfaceNormalOffset);
+                
+                // Draw the normal of the raycast hit for debugging 
+                Debug.DrawRay(currentHit.point, currentHit.normal * 0.5f, Color.yellow);
+            }
+            else
+            {
+                GoalPosition = SolverHandler.TransformTarget.position + (SolverHandler.TransformTarget.forward * DefaultPlacementDistance);
+            }
+        }
+
+        protected virtual void SetRotation()
+        {
+            Vector3 direction = currentRay.Direction;
+            Vector3 surfaceNormal = currentHit.normal;
+
+            if (KeepOrientationVertical)
+            {
+                direction.y = 0;
+                surfaceNormal.y = 0;
+            }
+
+            // if the object is on a surface then change the rotation according to the normal of the hit point
+            if (DidHit && rotateAccordingToSurface)
+            {
+                GoalRotation = Quaternion.LookRotation(-surfaceNormal, Vector3.up);
+            }
+            else 
+            {
+                // If (DidHit and !rotateAccordingToSurface) or if there is no raycast hit, 
+                GoalRotation = Quaternion.LookRotation(direction, Vector3.up);
+            }
+        }
+
+        private void ToggleSpatialMeshVisibility(bool spatialMeshVisibility)
+        {
+            // If the spatial mesh is already visible on start, then we do not need to toggle the spatial mesh 
+            // if the object is in the placing state
+            if (InitialSpatialMeshDisplayOption == SpatialAwarenessMeshDisplayOptions.None || InitialSpatialMeshDisplayOption == SpatialAwarenessMeshDisplayOptions.Occlusion)
+            {
+                IReadOnlyList<IMixedRealitySpatialAwarenessMeshObserver> observers = spatialAwarenessSystemAcecess.GetDataProviders<IMixedRealitySpatialAwarenessMeshObserver>();
```

Comment:
Look at using CoreServices.GetSPatialAwarenessDataProvider() api

Reply:
Removed SA dependency and added new events